### PR TITLE
param improvements

### PR DESCRIPTION
**Pull request by dagar**
nsh> perf
	...
	param_set: 0 events, 0us elapsed, 0us avg, min 0us max 0us 0.000us rms
	param_get: 5266 events, 55534us elapsed, 10us avg, min 4us max 1893us 57.558us rms
	param_export: 0 events, 0us elapsed, 0us avg, min 0us max 0us 0.000us rms

	nsh> param set SYS_AUTOSTART 2100
	nsh> param save

	nsh> perf
	...
	param_set: 0 events, 0us elapsed, 0us avg, min 0us max 0us 0.000us rms
	param_get: 6960 events, 78366us elapsed, 11us avg, min 4us max 1896us 61.081us rms
	param_export: 1 events, 1012771us elapsed, 1012771us avg, min 1012771us max 1012771us   NaNus rms

	nsh> param set SYS_MC_EST_GROUP 1
	  SYS_MC_EST_GROUP: curr: 2 -> new: 1
	nsh> param save

	nsh> perf
	...
	param_set: 1 events, 86us elapsed, 86us avg, min 86us max 86us Infinityus rms
	param_get: 18022 events, 196162us elapsed, 10us avg, min 4us max 2690us 62.364us rms
	param_export: 3 events, 3288253us elapsed, 1096084us avg, min 1012771us max 1247173us 131077.047us rms

### CODE REVIEW

**Comment by bkueng** [comment]
I'm wondering if it's worth adding the overhead to `param_get`, since it's a fast method (thus adding `perf_end` will have a noticeable impact) and it's called many times on a param change event.
Do we know what the cost is for a perf counter?

**Comment by LorenzMeier** [comment]
The overhead is not insignificant, that’s correct. Should this be a compile time option for benchmarks?

**Comment by dagar** [comment]
That could be fine, I mainly want to use this to find the abusers. There shouldn't be anywhere near this many calls to param_get().

**Comment by dagar** [comment]
On second thought I'd like to do some testing to see if we can tolerate leaving this on to catch misuse in regular operation. There are also more areas of the system where we shouldn't even be updating params while flying.

**Comment by dagar** [comment]
@bkueng could you give this an initial review (the last commit in particular). It drastically reduces the param export time on a pixhawk. The param export is called from either the param command `param save` or in LPWORK. In both cases there's plenty of stack available for the buffer.

Pixhawk 2 baseline

    param_export: 3 events, 3288253us elapsed, 1096084us avg, min 1012771us max 1247173us 131077.047us rms

Pixhawk 2 with bson encoder buffer

    param_export: 3 events, 30897us elapsed, 10299us avg, min 10219us max 10349us 70.004us rms

**Comment by bkueng** [comment]
Nice! I just glanced over it but it makes sense. So it seems that NuttX does not use a write buffer for the params file. I have a few questions/remarks:
- What is the contribution of `usleep(5);` vs. using the buffer alone?
- is there any change in error behavior when an export is interrupted? e.g. a power loss.
- there are a few more callers through `param_save_default()` in the codebase

**Comment by dagar** [comment]
> What is the contribution of usleep(5); vs. using the buffer alone?
I'm not sure exactly (it was dropped later on). I can do a quick test with and without. Do you see any reason to keep it? I believe all the potentially callers are low priority.

> is there any change in error behavior when an export is interrupted? e.g. a power loss.
Change yes, but I think either way you're screwed and end up with partial bson records and no bson end. If this is a major concern I have some other (initially unrelated) ideas for the param system that would actually help here.

> there are a few more callers through param_save_default() in the codebase
Thanks, I'll take a closer look. Even with a smaller buffer it's still worth it.


I've been testing across a random assortment of boards. Obviously the timing is dependant on the number of params, but curiously the pixhawk 2 cube seems to take much longer to save (often > 1s in master) compared to a pixracer (usually < 500ms).

**Comment by bkueng** [comment]
> I can do a quick test with and without. Do you see any reason to keep it?

No, I just like to know the individual contributing factors.

> If this is a major concern I have some other (initially unrelated) ideas for the param system that would actually help here.

Ok. So do I, but I think we're fine. Since it's much faster now, it's becoming less and less of a concern. And since the plan is for future boards to have a small backup battery, it will become easier to handle as well.

Did you see the NuttX options `CONFIG_MTD_WRBUFFER`, `CONFIG_MTD_NWRBLOCKS` and `CONFIG_RAMTRON_WRITEWAIT` (fyi @davids5 )?

**Comment by dagar** [comment]
I haven't looked experimented with the NuttX write buffers yet, but I'm wondering if we should go with this regardless. Even on LInux (desktop x86 or Raspi Arm) it's much faster.

master + perf
	param_export: 6 events, 15519us elapsed, 2586us avg, min 2281us max 3157us 318.050us rms

PR with bson encoder buffer
	param_export: 5 events, 116us elapsed, 23us avg, min 0us max 45us 21.534us rms

File: src/drivers/px4fmu/fmu.cpp [hunk 1]
```diff
@@ -1413,7 +1413,11 @@ PX4FMU::cycle()
 #endif
 		/* check arming state */
 		bool updated = false;
-		orb_check(_armed_sub, &updated);
+
+		// only update params if disarmed
+		if (!_armed.armed) {
+			orb_check(_armed_sub, &updated);
```
Comment (LorenzMeier) [inline src/drivers/px4fmu/fmu.cpp]: This looks potentially wrong - why are we only updating the armed state if we are disarmed? Generally we want PWM_REV and PWM_TRIM params to update also when armed, so if unsure this should be just dropped entirely.

**Comment by LorenzMeier** [comment]
Let's make sure to get this in quickly as the param export system load is a nuisance.

File: src/drivers/px4fmu/fmu.cpp (hunk 1 above)
Comment (dagar) [reply src/drivers/px4fmu/fmu.cpp]: The idea was to skip changing PWM parameters when flying, which could easily be unsafe (and a small performance optimization).

**Comment by LorenzMeier** [comment]
Good to go!

**Pull request merged**